ledger/src/blockstore.rs

```diff
@@ -1493,22 +1525,94 @@ impl Blockstore {
         shred_index < slot_meta.consumed || data_index.contains(shred_index)
     }
 
-    fn get_data_shred_from_just_inserted_or_db<'a>(
+    fn get_shred_from_just_inserted_or_db<'a>(
         &'a self,
         just_inserted_shreds: &'a HashMap<ShredId, Shred>,
         slot: Slot,
-        index: u64,
+        index: u32,
+        shred_type: ShredType,
     ) -> Cow<'a, Vec<u8>> {
-        let key = ShredId::new(slot, u32::try_from(index).unwrap(), ShredType::Data);
+        let key = ShredId::new(slot, index, shred_type);
         if let Some(shred) = just_inserted_shreds.get(&key) {
             Cow::Borrowed(shred.payload())
-        } else {
+        } else if shred_type == ShredType::Data {
             // If it doesn't exist in the just inserted set, it must exist in
             // the backing store
-            Cow::Owned(self.get_data_shred(slot, index).unwrap().unwrap())
+            Cow::Owned(
+                self.get_data_shred(slot, u64::from(index))
+                    .unwrap()
+                    .unwrap_or_else(|| {
+                        panic!("{} {} {:?} must be present!", slot, index, shred_type)
+                    }),
+            )
+        } else {
+            Cow::Owned(
+                self.get_coding_shred(slot, u64::from(index))
+                    .unwrap()
+                    .unwrap_or_else(|| {
+                        panic!("{} {} {:?} must be present!", slot, index, shred_type)
```

Comment:
What ensures that the shred is present here?
This is a pretty risky method that internally relies on certain presumptions where the caller may not be aware of.
I do see the logic was already present before this commit, but the commit is extending the uses of this method.

Reply:
Nothing. This method is used for cases where a meta structure has already informed you of the shreds existence, where it is preferable to panic if there is an inconsistency. 

However if you prefer I can return an option, and have the callers `error!` and continue instead.

Reply:
yeah, I think returning an `Option` is better here. Maybe in a separate pr first if that is changing too much of existing code.
this function as it stands is a pretty risky code.